mlos_core/mlos_core/optimizers/observations.py

```diff
@@ -0,0 +1,361 @@
+            ), "config and score must have the same index"
+            return
+
+        self.config = pd.concat([self.config, config]).reset_index(drop=True)
+        self.score = pd.concat([self.score, score]).reset_index(drop=True)
+        assert set(self.config.index) == set(
+            self.score.index
+        ), "config and score must have the same index"
+
+        if self.context is not None:
+            assert context is not None, (
+                "context of appending observation must not be null"
+                + " if context of prior observation is not null"
+            )
+            self.context = pd.concat([self.context, context]).reset_index(drop=True)
+            assert self.config.index.equals(
+                self.context.index
+            ), "config and context must have the same index"
+        else:
+            assert context is None, (
+                "context of appending observation must be null"
+                + " if context of prior observation is null"
+            )
+        if self.metadata is not None:
+            assert metadata is not None, (
+                "context of appending observation must not be null"
+                + " if metadata of prior observation is not null"
+            )
+            self.metadata = pd.concat([self.metadata, metadata]).reset_index(drop=True)
+            assert self.config.index.equals(
+                self.metadata.index
+            ), "config and metadata must have the same index"
+        else:
+            assert metadata is None, (
+                "context of appending observation must be null"
+                + " if metadata of prior observation is null"
+            )
+
+    def to_list(self) -> List[Observation]:
+        """
+        Converts the observations to a list of observations.
+
+        Returns
+        -------
+        List[Observation]
+            The list of observations.
+        """
+        return [
+            Observation(
+                config=self.config.loc[idx],
+                score=self.score.loc[idx],
+                context=None if self.context is None else self.context.loc[idx],
+                metadata=None if self.metadata is None else self.metadata.loc[idx],
+            )
+            for idx in self.config.index
+        ]
+
+    def __len__(self) -> int:
+        return len(self.config.index)
+
+    def __iter__(self) -> Iterator[Observations]:
+        for idx in self.config.index:
+            yield Observations(
+                config=self.config.loc[[idx]],
+                score=self.score.loc[[idx]],
+                context=None if self.context is None else self.context.loc[[idx]],
+                metadata=None if self.metadata is None else self.metadata.loc[[idx]],
+            )
+
+    def __repr__(self) -> str:
+        return (
+            f"Observation(config={self.config}, score={self.score},"
+            + " context={self.context}, metadata={self.metadata})"
```

Comment:
```suggestion
            " context={self.context}, metadata={self.metadata})"
```

Python doesn't need the + since the () wrap the expression already.

Reply:
I usually also like to leave the trailing space on the line before (nit).